### PR TITLE
go/consensus/tendermint: Add support for state sync

### LINKED ISSUE #2880: State sync for the Tendermint consensus backend
We should support state sync for the Tendermint consensus backend. Also requires at least basic support for consensus light clients (see #2440).

### PR DESCRIPTION
**Pull request by kostko**
Fixes #2880.
Based on #2881.

TODO
* [x] Cleanup.
  * [x] Make checkpoint parameters configurable (should likely be consensus parameters).
* [x] More tests.
* [x] Wait for Tendermint Core 0.34.

### CODE REVIEW

**Comment by kostko** [comment]
This is ready for early review, merge pending on Tendermint Core 0.34 release, but the "version bump commit" in this branch should be compatible with what will be in 0.34.

**Comment by kostko** [comment]
Rebased on top of 0.34.0-rc1 which was released today.

**Comment by kostko** [comment]
Our test suite found the following issue in 0.34.0-rc1: https://github.com/tendermint/tendermint/issues/5179 (happens both in `e2e/runtime/history-reindex` and `e2e/runtime/txsource`).

File: go/consensus/tendermint/light/client.go [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+	}
+
+	tmc, err := tmlight.NewClient(
+		cfg.GenesisDocument.ChainID,
+		cfg.TrustOptions,
+		providers[0],                       // Primary provider.
+		providers[1:],                      // Witnesses.
+		tmlightdb.New(tmdb.NewMemDB(), ""), // TODO: Make the database configurable.
+		tmlight.MaxRetryAttempts(5),        // TODO: Make this configurable.
+		// TODO: Refactor log adapter to tendermint common.
```
Comment (kostko) [inline go/consensus/tendermint/light/client.go:244]: This should be fixed once #3140 is merged.

File: go/consensus/tendermint/api/genesis.go [hunk 2]
```diff
@@ -72,19 +74,24 @@ func genesisToTendermint(d *genesis.Document) (*tmtypes.GenesisDoc, error) {
...
 				TimeIotaMs: 1000,
 			},
-			Evidence: tmtypes.EvidenceParams{
-				MaxAgeNumBlocks: int64(d.Consensus.Parameters.MaxEvidenceAgeBlocks),
-				MaxAgeDuration:  d.Consensus.Parameters.MaxEvidenceAgeTime,
+			Evidence: tmproto.EvidenceParams{
+				MaxAgeNumBlocks:  int64(d.Consensus.Parameters.MaxEvidenceAgeBlocks),
+				MaxAgeDuration:   d.Consensus.Parameters.MaxEvidenceAgeTime,
+				MaxNum:           50,
+				ProofTrialPeriod: 50_000,
```
Comment (kostko) [inline go/consensus/tendermint/api/genesis.go:87]: These should probably be transferred over to our general consensus parameters.

File: go/consensus/tendermint/tendermint.go [hunk 3]
```diff
@@ -1615,6 +1665,13 @@ func init() {
 	Flags.Bool(CfgDebugDisableCheckTx, false, "do not perform CheckTx on incoming transactions (UNSAFE)")
 	Flags.Bool(CfgDebugUnsafeReplayRecoverCorruptedWAL, false, "Enable automatic recovery from corrupted WAL during replay (UNSAFE).")
 
+	// State sync.
+	Flags.Bool(CfgConsensusStateSyncEnabled, false, "enable state sync")
+	Flags.StringSlice(CfgConsensusStateSyncConsensusNode, []string{}, "consensus node to use for syncing the light client")
+	Flags.Duration(CfgConsensusStateSyncTrustPeriod, 24*time.Hour, "light client trust period")
+	Flags.Uint64(CfgConsensusStateSyncTrustHeight, 0, "light client trusted height")
+	Flags.String(CfgConsensusStateSyncTrustHash, "", "light client trusted consensus header hash")
```
Comment (kostko) [inline go/consensus/tendermint/tendermint.go:1673]: Update these comments to refer specifically to state sync as there will later be a similar set of options for the light node.

File: go/consensus/tendermint/full/light.go [hunk 4]
```diff
@@ -54,9 +50,18 @@ func (t *fullService) GetValidatorSet(ctx context.Context, height int64) (*conse
 		return nil, consensusAPI.ErrVersionNotFound
 	}
 
+	protoVals, err := vals.ToProto()
+	if err != nil {
+		return nil, fmt.Errorf("tendermint: failed to marshal validators: %w", err)
+	}
+	meta, err := protoVals.Marshal()
+	if err != nil {
+		return nil, fmt.Errorf("tendermint: failed to marshal validators: %w", err)
```
Comment (Yawning) [inline go/consensus/tendermint/full/light.go:59]: Maybe distinguish from the previous error case somehow so I don't need to hunt down which line number it was on?

File: go/consensus/tendermint/abci/state.go [hunk 5]
```diff
@@ -252,6 +252,24 @@ func (s *applicationState) doInitChain(now time.Time) error {
...
+	defer s.blockLock.Unlock()
+
+	s.stateRoot = root
+
+	s.deliverTxTree.Close()
+	s.deliverTxTree = mkvs.NewWithRoot(nil, s.storage.NodeDB(), root, mkvs.WithoutWriteLog())
+	s.checkTxTree.Close()
+	s.checkTxTree = mkvs.NewWithRoot(nil, s.storage.NodeDB(), root, mkvs.WithoutWriteLog())
+
+	if err := s.doCommitOrInitChainLocked(time.Time{}); err != nil {
```
Comment (Yawning) [inline go/consensus/tendermint/abci/state.go:266]: Ensure nothing bad will happen due to the use of `time.Time{}`.

Comment (kostko) [reply go/consensus/tendermint/abci/state.go:266]: This time would only be used for transaction simulation (for gas estimation) until the first new block is committed. This is the same as currently happens on initial genesis until the first block is committed.

File: go/go.mod [hunk 6]
```diff
@@ -7,7 +7,7 @@ replace (
 	// https://github.com/spf13/cobra/issues/1091
 	github.com/gorilla/websocket => github.com/gorilla/websocket v1.4.2
 
-	github.com/tendermint/tendermint => github.com/oasisprotocol/tendermint v0.33.6-oasis1
+	github.com/tendermint/tendermint => github.com/oasisprotocol/tendermint v0.32.0-dev2.0.20200803174517-82f764f634eb
```
Comment (Yawning) [inline go/go.mod:10]: Whowa, that's an odd branch to be pointing this at.

Comment (kostko) [reply go/go.mod:10]: That's because there is no tag yet, only a branch. Will make a tag now and fix this.

Comment (kostko) [reply go/go.mod]: Fixed.

**Pull request merged**